src/python/WMComponent/WorkflowUpdater/WorkflowUpdaterPoller.py

```diff
@@ -0,0 +1,169 @@
+
+    def getPileupDocs(self):
+        """
+        Fetch all pileup documents from MSPileup and preprocess the data.
+        :return: a list of dictionaries in the following format:
+          {"pileupName": string with pileup name,
+           "customName": string with custom pileup name - if any,
+           "rses": list of RSE names,
+           "blocks": list of block names}
+        """
+        mgr = RequestHandler()
+        headers = {'Content-Type': 'application/json'}
+        data = mgr.getdata(self.msPileupUrl, params={}, headers=headers, verb='GET',
+                           ckey=self.userKey, cert=self.userCert, encode=True, decode=True)
+        if data and data.get("result", []):
+            if "error" in data["result"][0]:
+                msg = f"Failed to retrieve MSPileup documents. Error: {data}"
+                raise WorkflowUpdaterException(msg)
+
+        logging.info("A total of %d pileup documents have been retrieved.", len(data["result"]))
+        pileupMapList = []
+        for puItem in data["result"]:
+            logging.info("Pileup: %s, custom name: %s, expected at: %s, but currently available at: %s",
+                         puItem['pileupName'], puItem['customName'],
+                         puItem['expectedRSEs'], puItem['currentRSEs'])
+            thisPU = {"pileupName": puItem['pileupName'],
+                      "customName": puItem['customName'],
+                      "rses": puItem['currentRSEs'],
+                      "blocks": []}
+            pileupMapList.append(thisPU)
+        return pileupMapList
+
+    def findWflowsWithPileup(self, listSpecs):
+        """
+        Given a list of workflow names and their respective specs, load each
+        one of them and filter out those that don't require any pileup dataset.
+        :param listSpecs: a list of dictionary with workflow name and spec path
+        :return: a list with the workflow names that require pileup
+        """
+        wflowsWithPU = []
+        for wflowSpec in listSpecs:
+            try:
+                workloadHelper = WMWorkloadHelper()
+                workloadHelper.load(wflowSpec['spec'])
+                pileupSpecs = workloadHelper.listPileupDatasets()
+                if pileupSpecs:
+                    logging.info("Workflow: %s requires pileup dataset(s): %s",
+                                 wflowSpec['name'], pileupSpecs.values())
+                    wflowsWithPU.append(wflowSpec['name'])
+                else:
+                    logging.info("Workflow: %s does not require any pileup", wflowSpec['name'])
+            except Exception as ex:
+                msg = f"Failed to load spec file for: {wflowSpec['spec']}. Details: {str(ex)}"
+                logging.error(msg)
+        return wflowsWithPU
+
+    def findRucioBlocks(self, uniquePUList, pileupMap):
+        """
+        Given a list of unique pileup dataset names, list all of
+        their blocks in Rucio
```

Comment:
I suggest to add appropriate comment into doc string about potential slowness of this function especially with increasing number of processed workflows. Moreover, I suggest to add `@timeFunction` decorator to explicitly dump time spent in this function such that we we'll not be guessing but rather can look it up from the log after operations.